### PR TITLE
Add access control to temporary credentials endpoints.

### PR DESCRIPTION
**Pull request by creechy**
**PR Checklist**

- [ ] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

This is a follow on to the overall access control feature to add access control to the credential vending endpoints. The access control rules for credential vending follows the same rules as getting the entity.

This resolve Issue https://github.com/unitycatalog/unitycatalog/issues/501

### CODE REVIEW

File: server/src/main/java/io/unitycatalog/server/service/TemporaryModelVersionCredentialsService.java [hunk 1]
```diff
@@ -29,7 +36,19 @@ public TemporaryModelVersionCredentialsService(CredentialOperations credentialOp
     }
 
     @Post("")
+    @AuthorizeExpression("""
```
Comment (kriscon-db) [inline server/src/main/java/io/unitycatalog/server/service/TemporaryModelVersionCredentialsService.java:39]: So to match how DB UC handles this, we need to apply the FunctionExecute auth action when the requested credential is READ, and apply FunctionUpdate auth action when the requested credential is a READ/WRITE.

Comment (kriscon-db) [reply server/src/main/java/io/unitycatalog/server/service/TemporaryModelVersionCredentialsService.java:39]: See: https://src.dev.databricks.com/databricks/universe@ae2dac8a44b0e98d8ff18042bcf4a626476e36ad/-/blob/managed-catalog/src/main/scala/com/databricks/backend/managedcatalog/rpc/core/credential/GenerateTemporaryModelVersionCredentialHandler.scala?L252-256

File: server/src/main/java/io/unitycatalog/server/service/TemporaryModelVersionCredentialsService.java [hunk 2]
```diff
@@ -66,4 +79,35 @@ private Set<CredentialContext.Privilege> modelVersionOperationToPrivileges(Model
...
+    private void authorizeForOperation(GenerateTemporaryModelVersionCredential generateTemporaryModelVersionCredentials) {
+
+        // TODO: This is a short term solution to conditional expression evaluation based on additional request parameters.
+        // This should be replaced with more direct annotations and syntax in the future.
+
+        String readExpression = """
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorize(#principal, #catalog, OWNER) ||
+          (#authorize(#principal, #catalog, USE_CATALOG) && #authorize(#principal, #schema, OWNER)) ||
+          (#authorizeAny(#principal, #registered_model, OWNER, EXECUTE) && #authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG))
```
Comment (rameshchandra) [inline server/src/main/java/io/unitycatalog/server/service/TemporaryModelVersionCredentialsService.java:92]:
```suggestion
          #authorizeAny(#principal, #registered_model, OWNER, EXECUTE) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG)
```

File: server/src/main/java/io/unitycatalog/server/service/TemporaryTableCredentialsService.java [hunk 3]
```diff
@@ -29,8 +33,15 @@ public TemporaryTableCredentialsService(CredentialOperations credentialOps) {
   }
 
   @Post("")
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorize(#principal, #catalog, OWNER) ||
+          (#authorize(#principal, #schema, OWNER) && #authorize(#principal, #catalog, USE_CATALOG)) ||
+          (#authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG) && #authorizeAny(#principal, #table, OWNER, SELECT))
```
Comment (rameshchandra) [inline server/src/main/java/io/unitycatalog/server/service/TemporaryTableCredentialsService.java:40]: Similar to model temp credentials, the required privileges for table credentials depend on the operations -- select for read privileges and select+modify for write privileges. The rules are:

For read operation:
```
#authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) && #authorize(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #table, OWNER, SELECT)
```

For read/write:
```
#authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) && 
#authorize(#principal, #catalog, OWNER, USE_CATALOG) && 
(#authorize(#principal, #table, OWNER) || #authorizeAll(#principal, #table, SELECT, MODIFY))
```

File: server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java [hunk 4]
```diff
@@ -51,4 +64,34 @@ private Set<CredentialContext.Privilege> volumeOperationToPrivileges(VolumeOpera
...
+  private void authorizeForOperation(GenerateTemporaryVolumeCredential generateTemporaryVolumeCredential) {
+
+    // TODO: This is a short term solution to conditional expression evaluation based on additional request parameters.
+    // This should be replaced with more direct annotations and syntax in the future.
+
+    String readExpression = """
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorize(#principal, #catalog, OWNER) ||
+          (#authorize(#principal, #schema, OWNER) && #authorize(#principal, #catalog, USE_CATALOG)) ||
+          (#authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG) && #authorizeAny(#principal, #volume, OWNER, READ_VOLUME))
```
Comment (rameshchandra) [inline server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java:77]:
```suggestion
          #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #volume, OWNER, READ_VOLUME)
```

File: server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java [hunk 5]
```diff
@@ -51,4 +64,34 @@ private Set<CredentialContext.Privilege> volumeOperationToPrivileges(VolumeOpera
...
+          """;
+
+    String writeExpression = """
+          (#authorize(#principal, #volume, OWNER) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA))
```
Comment (rameshchandra) [inline server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java:81]:
```suggestion
          #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && 
          #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) &&
          (#authorize(#principal, #schema, OWNER) || #authorizeAll(#principal, #volume, READ_VOLUME, WRITE_VOLUME))
```

**Review by kriscon-db — APPROVED** [review]
LGTM from the models side.

File: server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java [hunk 6]
```diff
@@ -51,4 +64,34 @@ private Set<CredentialContext.Privilege> volumeOperationToPrivileges(VolumeOpera
...
+
+    String readExpression = """
+          #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #volume, OWNER, READ_VOLUME)
+          """;
+
+    // TODO: add WRITE_VOLUME to the expression
+    String writeExpression = """
+          #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) &&
+          #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) &&
+          (#authorize(#principal, #schema, OWNER) || #authorizeAll(#principal, #volume, READ_VOLUME))
```
Comment (rameshchandra) [inline server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java:81]:
```suggestion
          #authorize(#principal, #schema, OWNER)
```

**Review by rameshchandra — APPROVED** [review]
LGTM. Please incorporate the write authorization for volume creds before merging.

File: server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java (hunk 6 above)
Comment (creechy) [reply server/src/main/java/io/unitycatalog/server/service/TemporaryVolumeCredentialsService.java:81]: This should be `#volume` not `#schema` right?

**Pull request merged**